---
title: "Adipokine EWAS - Script 18: Triangulation analysis for Adi CpGs"
author: "Lucy Sinke"
output: html_document
---


```{r}
library(tidyverse)
library(ggpubr)
library(ggrepel)
```

# Filter

Read in coloc and eQTM analyses to filter CpGs

```{r}
load('adipo_eqtm.Rdata')
```

```{r}
load("AdiTop.Rdata")
top <- top %>% dplyr::select(CPG=cpg, 
                             EWAS_BETA = base_meta_beta,
                             EWAS_SE = base_meta_se,
                             EWAS_P = base_meta_p,
                              EWAS_PADJ = base_meta_padj_fdr)
```

We want to look at 73 CpGs

```{r}
keep <- top$CPG
keep <- unique(keep)
```

# Observed effect

We need the mQTL-Adi effect for the top mQTL for each CpG:

## Extract top mQTLs

Load in the SMR results looking at DNAm to adiponectin, containing dat_full - SNP level results

```{r}
load("DNAm_to_Adi.Rdata")
head(dat_full)

dat_full <- dat_full %>% filter(exposure %in% keep) %>% 
  select(SNP, CPG = exposure, SNP_ADI = beta.outcome, SNP_CPG = beta.exposure) %>% 
  group_by(CPG) %>% 
  filter(SNP_ADI == max(SNP_ADI)) %>% 
  ungroup()

length(unique(dat_full$CPG))

LEFT <- dat_full
```


There are only 42 of these CpGs with results from MR (I guess some did not have mQTLs or values in Adiponectin GWAS)

Extract smallest p-value SNP for each CpG. If there are multiple use effect size.
Save beta.outcome (the effect of that SNP on IL-6)

```{r}
write_csv(LEFT, file="DNAmADI.csv")
```

```{r}
summary(LEFT)
```


The effect on adiponectin ranges from -0.04 to 0.02

# Predicted effect of DNAm on adiponectin

## mQTL effect on DNAm
 
This was also in dat_full. See above.

## DNAm effect on IL-6

Extract from EWAS

For each CpG join with base_meta_beta

So, the predicted effect should be the effect of +1 SNP allele on adiponectin via DNAm

What we have:

* The association of +1 SNP allele with DNAm (from GoDMC, so DNAm increases by GODMC_BETA)
* The association of +1 adiponectin with DNAm (from EWAS, so +1 adi associated with EWAS_BETA change in DNAm)

We need to convert the latter (EWAS_BETA) to the effect of a GODMC_BETA increase in DNAm on adiponectin

* +EWAS_BETA DNAm = +1 adi
* +1 DNAm = + 1/EWAS_BETA adi
* +GODMC_BETA DNAm = + GODMC_BETA/EWAS_BETA adi

So the predicted effect of +1 SNP allele on adiponectin via DNAm is GODMC_BETA/EWAS_BETA

```{r}
LEFT <- left_join(LEFT, top, by="CPG")
LEFT$PREDICTED <- LEFT$SNP_CPG / LEFT$EWAS_BETA
LEFT <- LEFT %>% filter(CPG %in% keep)
```

Check axis

```{r}
summary(LEFT)
```

```{r}
library(ggrepel)
library(ggpubr)

LEFT %>% filter(CPG %in% c('cg11851174', 'cg02235049'))
```


```{r}
png("tri_dnamADI.png", height=24, width=24, units="cm", res=300)
LEFT %>% 
  ggplot(aes(y=SNP_ADI, x=PREDICTED)) +
  geom_hline(aes(yintercept = 0), color='#6e7a90', size=1) +
  geom_vline(aes(xintercept = 0), color='#6e7a90', size=1) +
  geom_abline(slope=1, color='#6e7a90', linetype='dashed', size=1) +
  geom_point(color = ifelse(LEFT$CPG %in% c('cg11851174', 'cg02235049'), '#76c2e6', '#6e7a90'), size=5) +
  geom_text_repel(label=ifelse(LEFT$CPG == c('cg11851174', 'cg02235049') ,LEFT$CPG, ''),
            vjust=-2, hjust=1.5, color='#76c2e6', size=6) +
  stat_cor(method="pearson") +
  xlab("Predicted effect") + ylab("Observed effect") +
  ggtitle("DNAm cause of Adiponectin") +
  theme_bw() +
    theme(
    axis.text = element_text(
       size=18, 
       color="#3f4754"),
    axis.title = element_text(
      size=20, 
      hjust=0.5, 
      color="#3f4754"),
    panel.background = element_rect(
      fill="white"),
    panel.border = element_rect(
      color="#3f4754",
      fill=NA),
    panel.grid.major = element_line(
      color="grey95"),
    panel.grid.minor = element_line(
      color="grey95"),
    plot.background = element_rect(
      fill="white"))
dev.off()
```


***

# Observed effect of Adiponectin on DNAm

For this we need the PRS effect on each CpG
The PRS is composed of two SNPs: rs4537545 and rs6734238

```{r}
load("Adi_to_DNAm.Rdata")
dat <- dat %>% filter(outcome %in% keep)
```

PRS should be effect of SNP allele on CpG multiplied by SNP EAF and is the effect of adiponectin score on CpG

```{r}
PRS <- data.frame()
for(i in unique(dat$outcome)){
  dat_temp <- dat %>% filter(outcome == i)
  dat_temp$PRS_COMP <- dat_temp$beta.outcome * dat_temp$eaf.exposure
  dat_temp <- dat_temp %>% 
    select(SNP, PRS_COMP, CPG = outcome)
  prs_temp <- data.frame(PRS_CPG = sum(dat_temp$PRS_COMP), CPG = dat_temp$CPG[1])
  PRS <- rbind(PRS, prs_temp)
}

summary(PRS)
```

```{r}
RIGHT <- PRS
```

# Predicted effect

First we calculate the effect of the PRS on adiponectin (which is -0.035, higher than on CpGs as expected)

```{r}
RIGHT$PRS_ADI <- (dat$beta.exposure[1] * dat$eaf.exposure[1]) + (dat$beta.exposure[2] * dat$eaf.exposure[2]) +
  (dat$beta.exposure[3] * dat$eaf.exposure[3]) + (dat$beta.exposure[4] * dat$eaf.exposure[4]) + (dat$beta.exposure[5] * dat$eaf.exposure[5]) 
RIGHT
```

We then want the predicted effect (effect of PRS on CpG via adi)

* Effect of SNP allele +1 on adi (GWAS_BETA)
* Effect of GWAS_BETA increase in adi on CpG

What we have is the effect of +1 adi on CpG (EWAS_BETA)

* +1 adi associated with EWAS_BETA change in DNAm
* + GWAS_BETA adi associated with EWAS_BETA * GWAS_BETA change in DNAm

```{r}
RIGHT <- left_join(RIGHT, top, by="CPG")
RIGHT$PREDICTED <- RIGHT$PRS_ADI * RIGHT$EWAS_BETA
summary(RIGHT)
```


```{r}
png("tri_ADIdnam.png", height=24, width=24, units="cm", res=300)
RIGHT %>% 
  ggplot(aes(y=PRS_CPG, x=PREDICTED)) +
  geom_hline(aes(yintercept = 0), color='#6e7a90', size=1) +
  geom_vline(aes(xintercept = 0), color='#6e7a90', size=1) +
  geom_abline(slope=1, color='#6e7a90', linetype='dashed', size=1) +
  geom_point(color = '#6e7a90', size=5) +
  stat_cor(method="pearson") +
  xlab("Predicted effect") + ylab("Observed effect") +
  ggtitle("DNAm cause of Adiponectin") + 
  theme_bw() +
    theme(
    axis.text = element_text(
       size=18, 
       color="#3f4754"),
    axis.title = element_text(
      size=20, 
      hjust=0.5, 
      color="#3f4754"),
    panel.background = element_rect(
      fill="white"),
    panel.border = element_rect(
      color="#3f4754",
      fill=NA),
    panel.grid.major = element_line(
      color="grey95"),
    panel.grid.minor = element_line(
      color="grey95"),
    plot.background = element_rect(
      fill="white"))
dev.off()
```

* Observed effect ranges from -0.014 to 0.007
* Predicted range

```{r}
write_csv(RIGHT, file="ADIDNAm.csv")
```




